File: zika_code/branch.R
rm(list=ls())
if(grepl('spencerfox', Sys.info()['login'])) setwd('~/projects/zika_alarm/zika_code/')
if(grepl('vagrant', Sys.info()['user'])) setwd('/vagrant/zika_alarm/zika_code/')
if(grepl('sjf826', Sys.info()['login'])) setwd('/home1/02958/sjf826/zika_alarm/zika_code/')
if(grepl('tacc', Sys.info()['nodename'])) setwd('/home1/02958/sjf826/zika_alarm/zika_code/')
if(grepl('meyerslab', Sys.info()['login'])) setwd('~/Documents/zika_alarm/zika_code/')
if(grepl('laurencastro', Sys.info()['login'])) setwd('~/Documents/zika_alarm/zika_code/')

sapply(c('branch.functions.R','plot.functions.R'), source)
library(plyr)
library(cowplot)
# prop_p -- probability an I infects a new individual in a time period: how to determine this? 
# recov_p -- probability an I recovers in a time period: human recovery 
# disc_p -- probability we discover an I: symptom ratio 
# d_thresh -- unused... number of discoveries
# e_thresh -- total instantaneous I's that count as epidemic escape (not cumulative):

#Parameters 
branch_params <- function(prop_p = 1.7/7 , 
                          recov_p = 1.0/7,
                          d_thres = 5,
                          e_thresh = 300,
                          prob_symp = 1,
                          incub_p = 1,
                          dis_prob_symp = 1,
                          dis_prob_asymp = 0.00 ,
                          intro_rate = 0.000)
  return(as.list(environment()))

# Average Run 
trials <- run_branches(num_reps = 1000, branch_params())
lastvalues <- all_last_cuminfect_values(trials)
hist(lastvalues, breaks = 100)
hist(lastvalues, breaks = 100, ylim = c(0,100))
sort(lastvalues)

# R0 = 1 Run 
trials2 <- run_branches(num_reps = 1000, branch_params(prop_p=.5/7))
lastvalues2 <- all_last_cuminfect_values(trials2)
hist(lastvalues2, breaks = 100)
hist(lastvalues2, breaks = 100, ylim = c(0,100))
sort(lastvalues2)


count_escapes(trials, 1, branch_params()$e_thresh)

plot_final_sizes(trials)


escape_prob <- prob_ext(prop_p = prop_p,recov_p = recov_p, incub_p = incub_p, prob_symp = prob_symp, 
         d_thres=d_thres,e_thresh = e_thresh, dis_prob_asymp, dis_prob_symp, intro_rate, nsamples=100)
escape_prob

escape_underDoverE <- prob_underD.overE(prop_p = prop_p,recov_p = recov_p, incub_p = incub_p, prob_symp = prob_symp, 
                        d_thres,e_thresh = e_thresh, dis_prob_asymp, dis_prob_symp, intro_rate, nsamples=100)


# Given we have detected X cases- what is the average number of Total Infected Cases and Cumulative Cases 
#at.detect <- all_detect_rows(trials)


## Plot grid of histograms, not really of interest anymore
# dis_prob_symps <- seq(0.01, 1, length.out = 5)
# prop_ps <- seq(1.1/7, 4/7, length.out = 5)
# all_runs <- data.frame()
# for(dis_prob_symp in dis_prob_symps){
#   for(prop_p in prop_ps){
#     trials <- run_branches(num_reps = 1000, prop_p, recov_p, incub_p, prob_symp, 
#                            d_thres, e_thresh, dis_prob_asymp, dis_prob_symp, intro_rate)
#     all_runs <- rbind(all_runs, data.frame(dis_prob_symp=dis_prob_symp, prop_p=prop_p, final_size = all_last_cuminfect_values(trials)))
#   }
# }
# 
# grid_hist <- ggplot(all_runs, aes(final_size)) + geom_histogram(binwidth = 20) + 
#   facet_grid(dis_prob_symp~prop_p, labeller=label_both, scales = "free_y")+
#   scale_y_continuous(expand=c(0,0))+
#   scale_x_continuous(expand=c(0.01,0.01))
# 
# ggsave(filename = "../ExploratoryFigures/Final_size_hist.pdf",plot = grid_hist, width=16, height=10)

File: zika_code/branch.py
import scipy

# Three kinds of counters:
# UI - Undiscovered Infecteds
# DI - Discovered Infecteds
# D - Cumulative number of discoveries
#
# Simulate until we have either:
# - we run out of infecteds
# - e_thresh number of infecteds, and d_thresh number of discoveries
#
# prop_p -- probability an I infects a new individual in a time period
# recov_p -- probability an I recovers in a time period
# disc_p -- probability we discover an I
# d_thresh -- unused... number of discoveries
# e_thresh -- total instantaneous I's that count as epidemic escape (not cumulative)
def run_branch(prop_p, recov_p, disc_p, d_thresh, e_thresh):
    (UI,DI,D) = 1,0,0
    I = UI + DI
    time_record = [(I,D)]
    while ( (I < e_thresh) and (I > 0) ) or ( (I > 0) and (D < d_thresh) ):
        newDI_draws = scipy.rand(DI)
        newDI_count = scipy.sum( newDI_draws < prop_p) 
        decDI_draws = scipy.rand(DI)
        decDI_count = scipy.sum( decDI_draws < recov_p) 

        newUI_draws = scipy.rand(UI)
        newUI_count = scipy.sum( newUI_draws < prop_p) 
        decUI_draws = scipy.rand(UI)
        decUI_count = scipy.sum( decUI_draws < recov_p) 
        disUI_draws = scipy.rand(UI)
        disUI_count = scipy.sum( disUI_draws < disc_p) 

        removeUI = scipy.sum( (decUI_draws < recov_p) | ( disUI_draws < disc_p) )
        removeUDI = scipy.sum( (decUI_draws < recov_p) & ( disUI_draws < disc_p) )

        UI = UI + newUI_count + newDI_count - removeUI
        DI = DI - decDI_count + disUI_count - removeUDI
        D = D + disUI_count
        I = UI + DI
        time_record.append( (I,D) )
    return time_record

# Estimate the probability we will reach e_thresh number of I's (as opposed to
# the infection dying out) given we have discovered at least d_thresh cases
#
# prop_p -- probability an I infects a new individual in a time period
# recov_p -- probability an I recovers in a time period
# disc_p -- probability we discover an I
# d_thresh -- unused... number of discoveries
# e_thresh -- total instantaneous I's that count as epidemic escape (not cumulative)
def prob_ext(prop_p, recov_p, disc_p, d_thresh, e_thresh, nsamples=10000):
    escapes = 0
    i = 0
    while i < nsamples:
        record = run_branch(prop_p, recov_p, disc_p, d_thresh, e_thresh)
        I,D = record[-1]
        if D < d_thresh:
            continue
        if I > e_thresh:
            escapes += 1
        i += 1 
        print 'Estimate', float(escapes) / i

# R0 is prop_p / recov_p (propagation probability / recovery probability)
# disc_p is probability we discover a case
# disc_threshold is how many cases we have "discovered"
# and e_threshold is the escape number
# estimates the probability escape will happen, if we have discovered at least d_thresh cases
prob_ext(2.0/7, 1.0/7, .01, 5, 150)

